helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java

```diff
@@ -306,6 +317,26 @@ public ZKHelixManager(String clusterName, String instanceName, InstanceType inst
     }
     _monitorLevel = configuredMonitorLevel;
 
+    if (helixManagerProperty != null) {
+      HelixCloudProperty helixCloudProperty = _helixManagerProperty.getHelixCloudProperty();
+      if (helixCloudProperty != null && helixCloudProperty.isCloudEventCallbackEnabled()) {
+        // Load callback implementation class
+        if ((_cloudEventCallbackProperty = helixCloudProperty.getCloudEventCallbackProperty())
+            != null) {
+          _callbackImplClass = loadCloudEventCallbackImplClass(
+              _cloudEventCallbackProperty.getCallbackImplClassName());
+          LOG.info("Using {} as cloud event callback impl class.",
+              _callbackImplClass.getClass().getCanonicalName());
+        }
+        if (_callbackImplClass == null) {
+          throw new HelixException("No implementation class is loaded for cloud event callback.");
+        }
+
+        // Register this helix manager as a listener
+        CloudEventHandlerFactory.getInstance().registerCloudEventListener(this);
```

Comment:
So we are leaking "this" in a constructor. I'm not sure what it will mean for VM freeze case, but generally, it's potentially an issue where external class with this reference could call in another thread before "this" is fully constructed.

Reply:
Good point. Will look into this.

Reply:
The code is refactored, we will instantiate a listener when Helix manager connect() is called, so this issue is avoided.